interpreter/interpreter.go

```diff
@@ -148,89 +116,71 @@ func (p *packageObject) Equal(rhs values.Value) bool {
-}
-
-// Return gives the return value from the block
-func (itrp *Interpreter) Return() values.Value {
-	return itrp.globals.Return()
-}
-
-// GlobalScope returns a pointer to the global scope of the package.
-// That is the scope nested directly below the options scope.
-func (itrp *Interpreter) GlobalScope() *Scope {
-	return itrp.globals
-}
-
-// Option returns a Flux option by name
-func (itrp *Interpreter) Option(name string) values.Value {
-	return itrp.options.Get(name)
+	types     map[semantic.Node]semantic.Type
+	polyTypes map[semantic.Node]semantic.PolyType
+	vars      map[string]values.Value
+	vals      []values.Value
+	pkg       string
 }
 
-// SetOption sets a new option binding
-func (itrp *Interpreter) SetOption(name string, val values.Value) {
-	itrp.options.Set(name, val)
+func NewInterpreter() *Interpreter {
+	return &Interpreter{
+		types:     make(map[semantic.Node]semantic.Type),
+		polyTypes: make(map[semantic.Node]semantic.PolyType),
+		vars:      make(map[string]values.Value),
+	}
 }
 
-// SideEffects returns the evaluated expressions of a Flux package
-func (itrp *Interpreter) SideEffects() []values.Value {
-	return itrp.pkg.SideEffects()
+func (itrp *Interpreter) Clear() {
+	itrp.pkg = ""
+	for k := range itrp.vars {
+		delete(itrp.vars, k)
+	}
+	for n := range itrp.types {
+		delete(itrp.types, n)
+		delete(itrp.polyTypes, n)
+	}
+	itrp.vals = itrp.vals[:0]
 }
 
-// Package returns the current package object
+// Package constructs and returns a Flux package object from
+// an interpreter's global variables.
 func (itrp *Interpreter) Package() Package {
-	return itrp.pkg
+	exported := make(map[string]values.Value, len(itrp.vars))
+	types := make(map[string]semantic.Type, len(itrp.vars))
+	polyTypes := make(map[string]semantic.PolyType, len(itrp.vars))
+	for k, v := range itrp.vars {
+		exported[k] = v
+		types[k] = v.Type()
+		polyTypes[k] = v.PolyType()
+	}
+	return &packageObject{
+		name:        itrp.pkg,
+		exported:    exported,
+		types:       types,
+		polyTypes:   polyTypes,
+		sideEffects: itrp.vals,
+	}
 }
 
 // Eval evaluates the expressions composing a Flux package.
-func (itrp *Interpreter) Eval(pkg semantic.Node, importer Importer) error {
+func (itrp *Interpreter) Eval(node semantic.Node, scope Scope, importer Importer) error {
 	extern := &semantic.Extern{
-		Block:       &semantic.ExternBlock{Node: pkg},
-		Assignments: make([]*semantic.ExternalVariableAssignment, 0, itrp.globals.Len()+itrp.options.Len()),
+		Block: &semantic.ExternBlock{
+			Node: node,
+		},
+		Assignments: make([]*semantic.ExternalVariableAssignment, 0, scope.Size()),
 	}
-	// Add declarations for values in scope
-	addExternalAssignments(extern, itrp.options)
-	addExternalAssignments(extern, itrp.globals)
+
+	// TODO(jlapacik): Does it make sense to range over all variables currently in scope?
+	// Perhaps type inference should happen before evaluation as a separate step?
+	scope.Range(func(k string, v values.Value) {
```

Comment:
This seems good enough for now. We just need to tell the semantic inference logic that these names exist in scope and their types.